opm/parser/eclipse/EclipseState/Schedule/ScheduleState.hpp

```diff
@@ -54,6 +54,22 @@ namespace Opm {
 
     class WellTestConfig;
 
+    template <typename T>
+    class StateMember {
```

Comment:
Probably missing something here, but:
Is this wrapper really needed (or just making the code less readable)? It seems like the only reason `StateMember` is there, is for the update method (as `get()` is just dereferencing) which allows to create a new `shared_ptr` by moving an existing object. Couldn't that be done without the wrapper by a function
```
template<class T>
std::shared_ptr<T> updateState(T&& object){
   return std::make_shared<T>( std::move(object) );
}
...
updateState(this->snapshots.back().gconsale, std::move(new_gconsale) );
```
Thus everybody would still see that shared_ptrs are used and do not have to into Statemember.

Reply:
My goal has been to hide the `shared_ptr` from the API and make it easier to use - but I agree it looks a bit weird now. My suggestion is to add a `operator()` which does dereferencing - then downstream[1] can do:
```
const auto& gconsale = sched_state.gconsale();
```
instead of:
```
const auto& gconsale = sched_state.gconsale.get();
```

To me the most important is simple downstream usage; and I am willing to accept some funny stuff in opm-common to achieve that. The `StateMember` class and the `update` method is only relevant in the opm-common implementation. 

What do you think?

Reply:
I will stop arguing after this last comment:

My main point is, that every programmer knows how to handle a shared_ptr. With a wrapper I need to look up usage from the documentation or worse from the code. It will make working on the code harder for others.

Reply:
I fully understand your point. On the one hand there is value in making classes which closely match the use case, on the other hand it is valuable to stick to naked standard types like e.g. `std::shared_ptr` in this case. I think this is a quite classical conflict between the writer of the code who enjoys writing functionality which "exactly captures the problem" - and the reader who prefers well known datastructures like `std::something`. Personally I have shifted towards more use of naked `std::something` - but in this particular case I have ended on on the other side of the fence.

C++ wise there is of course only one API - but conceptually the code in question has two APIs, the downstream consumer API and the API in opm-common which is used during the assembly. In the current form the downstream API will be:
```C++
const auto& gconsump = sched_state.gconsump();
```
which I think is quite nice. In opm-common it is a bit more contrived of course. I'll add some comments.